pkg/credentialprovider/external_test.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package credentialprovider
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/registrycredentials"
+)
+
+func TestConfigDeserialization(t *testing.T) {
+	tests := []struct {
+		configYaml    []byte
+		expectedObj   *registrycredentials.RegistryCredentialConfig
+		matchExpected bool
+	}{
+		{
+			configYaml: []byte(`apiVersion: registrycredentials.k8s.io/v1alpha1
+kind: RegistryCredentialConfig
+providers:
+-
+  imageMatchers:
+  - "*.dkr.ecr.*.amazonaws.com"
+  - "*.dkr.ecr.*.amazonaws.com.cn"
+  exec:
+    command: ecr-creds
+    args:
+    - token
+    env:
+    - name: XYZ
+      value: envvalue
+    apiVersion: registrycredentials.k8s.io/v1alpha1`),
+			expectedObj: &registrycredentials.RegistryCredentialConfig{
+				Providers: []registrycredentials.RegistryCredentialProvider{
+					registrycredentials.RegistryCredentialProvider{
+						ImageMatchers: []string{
+							"*.dkr.ecr.*.amazonaws.com",
+							"*.dkr.ecr.*.amazonaws.com.cn",
+						},
+						Exec: registrycredentials.ExecConfig{
+							Command: "ecr-creds",
+							Args:    []string{"token"},
+							Env: []registrycredentials.ExecEnvVar{
+								registrycredentials.ExecEnvVar{
+									Name:  "XYZ",
+									Value: "envvalue",
+								},
+							},
+							APIVersion: "registrycredentials.k8s.io/v1alpha1",
+						},
+					},
+				},
+			},
+			matchExpected: true,
+		},
+	}
+	for _, test := range tests {
+		matched, err := testDecode(test.configYaml, test.expectedObj)
+		if err != nil {
+			t.Errorf("Test failed with error: %s", err.Error())
+		}
+		if matched != test.matchExpected {
+			t.Errorf("Expected match result to be %t, but was %t:\n=== Yaml:\n\n%s\n\n=== Object:\n\n%s\n===\n",
+				test.matchExpected, matched, test.configYaml, configString(test.expectedObj))
+		}
+	}
+}
+
+func testDecode(yaml []byte, expectedObj *registrycredentials.RegistryCredentialConfig) (bool, error) {
+	actualObj, err := decode(yaml)
+	if err != nil {
+		return false, err
+	}
+
+	if len(expectedObj.Providers) != len(actualObj.Providers) {
```

Comment:
Everything bellow this could be:
```
if diff := cmp.Diff(expectedObj, actualObj); diff != "" {
  t.Errorf("Unexpected diff (-want +got):\n%s", diff)
}
```

Reply:
Thanks, I'll give it a try.